rainier-core/src/main/scala/rainier/sampler/Variational.scala

```diff
@@ -0,0 +1,127 @@
+package rainier.sampler
+
+import rainier.compute._
+import rainier.core._
+
+trait Guide {
+  def density: Real // density of variational distributions
+  def params: Seq[Variable] // params to optimize
+  def forward: Map[Variable, Real] // map between original model params and f(params, eps)
+  def generate(implicit rng: RNG): Seq[Double]
+  def sample(paramSettings: Map[Variable, Double]): Map[Variable, Double]
+  def init: Map[Variable, Double]
+
+  private def transformDensity(targetDensity: Real): Real = {
+    Real.substituteVariable(targetDensity, forward)
+  }
+  def loss(targetDensity: Real): Real = {
+    transformDensity(targetDensity) - density
+  }
+}
+
+case class MeanFieldNormalGuide(targetVariables: Seq[Variable])(
+    implicit rng: RNG)
+    extends Guide {
+  private val normals: Map[(Variable, Variable), (Real, Continuous)] =
+    targetVariables
+      .map(tv => {
+        val mu = new Variable
+        val sigma = new Variable
+        val dist = Normal(mu, sigma)
+        ((mu, sigma), (dist.param.density.variables.head, dist))
+      })
+      .toMap
+
+  override def params: Seq[Variable] = normals.keys.toSeq.flatMap {
```

Comment:
you don't need `override` when the trait just declared the signature without a method body